src/stream/src/executor/backfill/no_shuffle_backfill.rs

```diff
@@ -233,118 +238,88 @@ where
+                            stream::PollNext::Left
+                        });
+
+                    #[for_await]
+                    for either in backfill_stream {
+                        match either {
+                            // Upstream
+                            Either::Left(msg) => {
+                                match msg? {
+                                    Message::Barrier(barrier) => {
+                                        // We have to process barrier outside of the loop.
+                                        // This is because the backfill stream holds a mutable
+                                        // reference to our chunk builder.
+                                        // We want to create another mutable reference
+                                        // to flush remaining chunks from the chunk builder
+                                        // on barrier.
+                                        // Hence we break here and process it after this block.
+                                        pending_barrier = Some(barrier);
+                                        break;
+                                    }
+                                    Message::Chunk(chunk) => {
+                                        // Buffer the upstream chunk.
+                                        upstream_chunk_buffer.push(chunk.compact());
+                                    }
+                                    Message::Watermark(_) => {
+                                        // Ignore watermark during backfill.
+                                    }
+                                }
+                            }
+                            // Snapshot read
+                            Either::Right(msg) => {
+                                match msg? {
+                                    None => {
+                                        // End of the snapshot read stream.
+                                        // We should not mark the chunk anymore,
+                                        // otherwise, we will ignore some rows
+                                        // in the buffer. Here we choose to never mark the chunk.
+                                        // Consume with the renaming stream buffer chunk without
+                                        // mark.
                                         for chunk in upstream_chunk_buffer.drain(..) {
-                                            cur_barrier_upstream_processed_rows +=
-                                                chunk.cardinality() as u64;
+                                            let chunk_cardinality = chunk.cardinality() as u64;
+                                            cur_barrier_snapshot_processed_rows +=
+                                                chunk_cardinality;
+                                            total_snapshot_processed_rows += chunk_cardinality;
```

Comment:
It seems we should use `cur_barrier_upstream_processed_rows += chunk_cardinality` instead.